producer/writer/message_writer.go

```diff
@@ -279,27 +290,41 @@ func (w *messageWriterImpl) retryUnacknowledged() {
 	var (
 		toBeRetried []*message
 		beforeRetry = w.nowFn()
+		batchSize   = w.opts.MessageRetryBatchSize()
 	)
 	for e != nil {
-		now := w.nowFn()
-		nowNanos := now.UnixNano()
+		beforeBatch := w.nowFn()
+		beforeBatchNanos := beforeBatch.UnixNano()
 		w.Lock()
-		e, toBeRetried = w.retryBatchWithLock(e, nowNanos)
+		e, toBeRetried = w.retryBatchWithLock(e, beforeBatchNanos, batchSize)
 		consumerWriters := w.consumerWriters
 		w.Unlock()
-		if len(consumerWriters) == 0 {
-			// Not expected in a healthy/valid placement.
-			w.m.noWritersError.Inc(int64(len(toBeRetried)))
-			w.m.retryBatchLatency.Record(w.nowFn().Sub(now))
-			continue
+		err := w.writeBatch(consumerWriters, toBeRetried)
+		w.m.retryBatchLatency.Record(w.nowFn().Sub(beforeBatch))
+		if err != nil {
+			// When we can't write to any consumer writer, skip the tick
+			// to avoid meaningless attempts, wait for next tick to retry.
+			break
 		}
+	}
+	w.m.retryTotalLatency.Record(w.nowFn().Sub(beforeRetry))
+}
 
-		for _, m := range toBeRetried {
-			w.write(consumerWriters, m)
+func (w *messageWriterImpl) writeBatch(
+	consumerWriters []consumerWriter,
+	toBeRetried []*message,
+) error {
+	if len(consumerWriters) == 0 {
+		// Not expected in a healthy/valid placement.
+		w.m.noWritersError.Inc(int64(len(toBeRetried)))
+		return errNoWriters
+	}
+	for _, m := range toBeRetried {
+		if err := w.write(consumerWriters, m); err != nil {
```

Comment:
If you fail to write one consumer writer, do you want to try the others? That seems to be the behavior before.

Reply:
yeah in the write() function it's still doing that behavior